Dockerfile.usability and host integration

christopherreay opened the pull request:
#### changes to `dockerfile.alpine` such that:
* gx and holochain are installed as root, which is much more docker friendly
* a new user is created every time an image instantiates into a container
  * the user has the username "user"
  * the uid of the new user is the same as the uid of the current user in the host
* have created the basic tools for a simple test harness
  * tests are run on the host
  * structure to run complex commands inside the *container* as part of the test
  * structure to pass information through to the host

#### creating container aware tests
* there are tests showing that hc will run as the user "user" inside a container, and serve the sample example
* `dockerfile.noCompile` adds the current filesystem into the latest build of the image without recompiling hc. This is to allow a development cycle of tests on top of a specific build

GerryG commented on tests/test.hc.serve.sample.test:
```diff
@@ -0,0 +1,4 @@
+docker run  -e LOCAL_USER_ID=`id -u $USER` -i devcore /bin/sh -c tests/test.hc.serve.sample.test.inContainer | tee test.hc.serve.sample.test.output
+grep -q "Starting with UID : 1002" test.hc.serve.sample.test.output || exit 1
+rm test.hc.serve.sample.test.output
+echo "passed test served hc as as user with matching user id"
```
I'm not sure I get the idea of creating a user in the container that matches the outside UID. This the test run user as distinct from the install owner (holochain)? I'm not sure there is any additional isolation gained.

GerryG replied:
What do you think of using $USER as the name inside too? Not really necessary but I think actually more clear than 'user'

christopherreay commented:
hi Gerry :)

I removed the "holochain" user, with a fixed id. I added links to the
rationale pages as comments in the Dockerfile.

gx and holochain are now installed as root, which suits the syntax design
of Dockerfile

hc is _run_ as a user inside the container which has the same uid as the
user that is running the container. this means that mounted filesystem,
specifically the .holochain directory are the same uid, which means files
can be effectively shared between contexts

this supports running tests inside the containers from the host, and
therefore test driven development of Dockerfile

On 21 Mar 2017 7:53 p.m., "Gerry Gleason" <notifications@github.com> wrote:

> *@GerryG* commented on this pull request.
> ------------------------------
>
> In tests/test.hc.serve.sample.test
> <https://github.com/metacurrency/holochain/pull/115#discussion_r107323472>
> :
>
> > @@ -0,0 +1,4 @@
> +docker run  -e LOCAL_USER_ID=`id -u $USER` -i devcore /bin/sh -c tests/test.hc.serve.sample.test.inContainer | tee test.hc.serve.sample.test.output
> +grep -q "Starting with UID : 1002" test.hc.serve.sample.test.output || exit 1
> +rm test.hc.serve.sample.test.output
> +echo "passed test served hc as as user with matching user id"
>
> I'm not sure I get the idea of creating a user in the container that
> matches the outside UID. This the test run user as distinct from the
> install owner (holochain)? I'm not sure there is any additional isolation
> gained.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/metacurrency/holochain/pull/115#pullrequestreview-28287115>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOnvjmtzZ-aT5w78QgtsWbLjKsSEiDhks5roI0igaJpZM4Mkp3R>
> .
>

GerryG commented:
I guess using mounted tests is a fairly good reason. I note that the volume feature can be a bit hinky, and in particular it is really slow on MacOS xhyve docker.

GerryG commented on Dockerfile.alpine:
```diff
@@ -9,40 +9,45 @@ RUN apk add --update \
+      && chmod 775 -R /work 
+
+# compile and test our development files
+  RUN make; make bs
```
In my view, tests should alway be run from outside the container, not at build time. Yes, do this line, but not the next. Making an image layer for the test run doesn't really make sense.

christopherreay replied:
Im looking at it from the perspective of a developer or a user who dont want something built if it doesnt makes sense for it to be built. assuming, from their perspective, the docker build isnt solely from the point of view of using hc

GerryG commented on Dockerfile.alpine:
```diff
@@ -9,40 +9,45 @@ RUN apk add --update \
-# above should get cached, does not depend on source
-ADD . /work/holochain 
+# make our development files available to hc, on top of the cached image "so far" from above
+  ADD . /work/holochain 
```
Didn't I see a different pull request with this as a COPY?

I would rename the entrypoint script, make it executable in the repo and run it from where it lands after it gets ADDed of COPYed in here. Then just add it to the RUN right below

christopherreay replied:
The ENTRYPOINT script timing for runnning is specific to its usecase I think

GerryG commented on Dockerfile.alpine:
```diff
@@ -9,40 +9,45 @@ RUN apk add --update \
+# sort out dynamic user login
+  COPY        entrypoint.sh.addHostUserToContainer /usr/local/bin/entrypoint.sh.addHostUserToContainer
+  RUN         chmod +x /usr/local/bin/entrypoint.sh.addHostUserToContainer
+  ENTRYPOINT  ["/usr/local/bin/entrypoint.sh.addHostUserToContainer"]
```
I would run this script, in a RUN command as commented above. No need to copy and chmod it, have it runable as installed above.

christopherreay replied:
Aye, I know, I dont get why the example was like that

GerryG commented on entrypoint.sh.addHostUserToContainer:
```diff
@@ -0,0 +1,16 @@
+# Either use the LOCAL_USER_ID if passed in at runtime or
+# fallback
+
+USER_ID=${LOCAL_USER_ID:-9001}
```
What do you think about RUN_USER_ID ? I assume you set in the run line, or we can do docker-compose.

I think there is a name we can give this script that will make it run at init time.

christopherreay replied:
the script does run after container construction, but before the CMD executes as per the specification of ENTRYPOINT

GerryG commented on entrypoint.sh.addHostUserToContainer:
```diff
@@ -0,0 +1,16 @@
+addgroup user sudoers
+su -c  "chown user:holochain /home/user -R"
+
+exec su-exec user "$@"
```
Now I get it. This magic will run what you supply on the run command line.

I guess with no args this will still give you a shell, otherwise the container stops when this script ends.
